plugins/linux/iptablesplugin/descriptor/rulechain.go

```diff
@@ -0,0 +1,418 @@
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at:
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package descriptor
+
+import (
+	"strings"
+
+	"github.com/gogo/protobuf/proto"
+	"github.com/pkg/errors"
+
+	"github.com/ligato/cn-infra/logging"
+	ifmodel "github.com/ligato/vpp-agent/api/models/linux/interfaces"
+	"github.com/ligato/vpp-agent/api/models/linux/iptables"
+	kvs "github.com/ligato/vpp-agent/plugins/kvscheduler/api"
+	ifdescriptor "github.com/ligato/vpp-agent/plugins/linux/ifplugin/descriptor"
+	"github.com/ligato/vpp-agent/plugins/linux/iptablesplugin/descriptor/adapter"
+	"github.com/ligato/vpp-agent/plugins/linux/iptablesplugin/linuxcalls"
+	"github.com/ligato/vpp-agent/plugins/linux/nsplugin"
+	nslinuxcalls "github.com/ligato/vpp-agent/plugins/linux/nsplugin/linuxcalls"
+)
+
+const (
+	// RuleChainDescriptorName is the name of the descriptor for Linux iptables rule chains.
+	RuleChainDescriptorName = "linux-ipt-rulechain-descriptor"
+
+	// dependency labels
+	ruleChainInterfaceDep = "interface-exists"
+
+	// minimum number of namespaces to be given to a single Go routine for processing
+	// in the Retrieve operation
+	minWorkForGoRoutine = 3
+)
+
+// A list of non-retriable errors:
+var (
+	// ErrCustomChainWithoutName is returned when the chain name is not provided for the custom iptables chain.
+	ErrCustomChainWithoutName = errors.New("iptables chain of type CUSTOM defined without chain name")
+
+	// ErrDefaultPolicyOnNonFilterRule is returned when a default policy is applied on a table different to FILTER.
+	ErrDefaultPolicyOnNonFilterRule = errors.New("iptables default policy can be only applied on FILTER tables")
+
+	// ErrDefaultPolicyOnCustomChain is returned when a default policy is applied on a custom chain, which is not allowed in iptables.
+	ErrDefaultPolicyOnCustomChain = errors.New("iptables default policy cannot be applied on custom chains")
+)
+
+// RuleChainDescriptor teaches KVScheduler how to configure Linux iptables rule chains.
+type RuleChainDescriptor struct {
+	log             logging.Logger
+	nsPlugin        nsplugin.API
+	scheduler       kvs.KVScheduler
+	ipTablesHandler linuxcalls.IPTablesAPI
+
+	// parallelization of the Retrieve operation
+	goRoutinesCnt int
+}
+
+// NewRuleChainDescriptor creates a new instance of the iptables RuleChain descriptor.
+func NewRuleChainDescriptor(
+	scheduler kvs.KVScheduler, ipTablesHandler linuxcalls.IPTablesAPI, nsPlugin nsplugin.API,
+	log logging.PluginLogger, goRoutinesCnt int) *RuleChainDescriptor {
+
```

Comment:
Recently we have come up with a cleaner layout for descriptor constructor: https://github.com/ligato/vpp-agent/blob/master/examples/kvscheduler/mock_plugins/ifplugin/descriptor/interface.go#L64-L92
It avoids the GetDescriptor method and registration inside the plugin is cleaner (plus it doesn't have to import adapter, which should be really only internal to descriptor): https://github.com/ligato/vpp-agent/blob/master/examples/kvscheduler/mock_plugins/ifplugin/ifplugin.go#L76-L81